### PR TITLE
Improved lightning error handling

### PR DESCRIPTION
**Pull request by cryptoquick**
- Updated with latest on development branch
- Use separate post_json_auth method
- Handles different response variants using this serde technique: https://serde.rs/enum-representations.html#untagged

### CODE REVIEW

File: src/operations/lightning.rs [hunk 1]
```diff
@@ -179,14 +181,31 @@ pub async fn pay_invoice(invoice: &str, token: &str) -> Result<String> {
...
-    // match serde_json::from_str::<PayInvoiceRes>(&response) {
-    //     Ok(response) => Ok(response),
-    //     Err(e) => Err(e),
-    // }
-    // let r: PayInvoiceError = serde_json::from_str(&response)?;
-
-    Ok(response)
+    let (response, _) = post_json_auth(&url, &Some(req), Some(token)).await?;
+
+    match serde_json::from_str::<PayInvoice>(&response) {
```
Comment (grunch) [inline src/operations/lightning.rs:186]: Thank you for your fix, returning the same type is a practical solution but not for all use cases, sometimes I want to get more than just the preimage.

I wanted to know if it's possible to return `PayInvoice {}` type if everything is ok and return the error `PayInvoice::Error { }` if there is something wrong

**Review by grunch — COMMENTED** [review]
Is it possible to remove all other commits? I want to have a clean branch with code changes only related with the lightning implementation, if it is not, no problem, I can create a new branch 😃

**Comment by cryptoquick** [comment]
> Is it possible to remove all other commits? I want to have a clean branch with code changes only related with the lightning implementation, if it is not, no problem, I can create a new branch smiley

If this is merged, the other commits will go away, since your PR is into development, and this merging in the latest from development.

File: src/operations/lightning.rs (hunk 1 above)
Comment (cryptoquick) [reply src/operations/lightning.rs:186]: It uses serde to parse untagged JSON responses into different enum variants, did you read the documentation I provided on that?
https://serde.rs/enum-representations.html#untagged
You can add other fields if you like.

Comment (grunch) [reply src/operations/lightning.rs:186]: > It uses serde to parse untagged JSON responses into different enum variants, did you read the documentation I provided on that? https://serde.rs/enum-representations.html#untagged You can add other fields if you like.

this is what I was looking for, thanks

**Comment by grunch** [comment]
> > Is it possible to remove all other commits? I want to have a clean branch with code changes only related with the lightning implementation, if it is not, no problem, I can create a new branch smiley
> 
> If this is merged, the other commits will go away, since your PR is into development, and this merging in the latest from development.

Yes but this makes difficult to be reviewed and keep track of what is happening with the PR

**Comment by cryptoquick** [comment]
> > > Is it possible to remove all other commits? I want to have a clean branch with code changes only related with the lightning implementation, if it is not, no problem, I can create a new branch smiley
> > 
> > 
> > If this is merged, the other commits will go away, since your PR is into development, and this merging in the latest from development.
> 
> Yes but this makes difficult to be reviewed and keep track of what is happening with the PR

I'll just merge it, then. That'll simplify things.

**Pull request merged**